sql/src/main/java/io/crate/metadata/Functions.java

```diff
@@ -21,27 +21,44 @@
 
 package io.crate.metadata;
 
+import com.google.common.base.Function;
 import com.google.common.base.Joiner;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
 import io.crate.metadata.tablefunctions.TableFunctionImplementation;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import io.crate.types.UndefinedType;
 import org.elasticsearch.common.inject.Inject;
 
 import javax.annotation.Nullable;
-import java.util.Locale;
-import java.util.Map;
+import java.util.*;
 
 public class Functions {
 
-    private final Map<FunctionIdent, FunctionImplementation> functionImplementations;
-    private final Map<String, DynamicFunctionResolver> functionResolvers;
+    private final Map<String, FunctionResolver> functionResolvers;
     private final Map<String, TableFunctionImplementation> tableFunctionImplementationMap;
 
     @Inject
     public Functions(Map<FunctionIdent, FunctionImplementation> functionImplementations,
-                     Map<String, DynamicFunctionResolver> functionResolvers,
+                     Map<String, FunctionResolver> functionResolvers,
                      Map<String, TableFunctionImplementation> tableFunctionImplementationMap) {
-        this.functionImplementations = functionImplementations;
-        this.functionResolvers = functionResolvers;
+        this.functionResolvers = Maps.newHashMap(functionResolvers);
         this.tableFunctionImplementationMap = tableFunctionImplementationMap;
+        generateFunctionResolvers(functionImplementations);
+    }
+
+    private void generateFunctionResolvers(Map<FunctionIdent, FunctionImplementation> functionImplementations) {
+
+        for (Map.Entry<FunctionIdent, FunctionImplementation> entry : functionImplementations.entrySet()) {
+            String functionName = entry.getKey().name();
+            GeneratedFunctionResolver resolver = (GeneratedFunctionResolver) functionResolvers.get(functionName);
+            if (resolver == null) {
+                resolver = new GeneratedFunctionResolver(functionName);
+                functionResolvers.put(functionName, resolver);
+            }
+            resolver.add(entry.getKey(), entry.getValue());
```

Comment:
would it be much work to create the map for the GeneratedFunctionResolver outside and instantiate the GeneratedFunctionResolver with the final map? I kinda dislike that the `signatures` method of the GeneratedFunctionResolver doesn't always return the same result - as it could change if `add` is used afterwards.